### PR TITLE
Add Windows CI

### PR DESCRIPTION
**Pull request by afonso360**
👋 Hey,

With the upgrade to Cranelift 0.88 we can now enable MSVC in the build script and add some CI. This also adds CI for MinGW on Windows, which I didn't expect to hit so many issues, but it turns out MinGW is even more broken than I thought.

A summary of issues / tests disabled:

* The JIT is still broken, so we disable all JIT tagged tests on windows
* `bench.simple-raytracer`
  * This is disabled due to hyperfine failing with an invalid preparation command
     Its just a build system issue.
* `test.regex-shootout-regex-dna`
  * This fails with a different output than expected
    The actual file is mostly the same, I think this is just different line endings
* `abi-cafe` patch
  * We've dropped the previous patch, in favour of the default config for disabling the i128 windows abi issues
  But it looks like its not broad enough since `x86_64-pc-windows-gnu` still fails, so we temporarily disable those
  I'm going to follow up with  `abi-cafe` to disable those tests on MinGW as well
* `rand` patch
  * There are a bunch of tests failing on rand, these only affect  `x86_64-pc-windows-gnu` presumably due to some broken implementations on the MinGW side? I haven't investigated this too far.

cc #977

### CODE REVIEW

File: patches/0001-rand-Disable-rand-tests-on-mingw.patch [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+diff --git a/rand_distr/src/pareto.rs b/rand_distr/src/pareto.rs
+index 217899ed9a..5f8e2c9bbb 100644
+--- a/rand_distr/src/pareto.rs
++++ b/rand_distr/src/pareto.rs
+@@ -107,6 +107,8 @@ mod tests {
+     }
+
+     #[test]
++    // This is broken on x86_64-pc-windows-gnu presumably due to a broken powf implementation
++    #[cfg(not(all(target_os = "windows", target_env = "gnu")))]
```
Comment (bjorn3) [inline patches/0001-rand-Disable-rand-tests-on-mingw.patch:20]: Maybe instead add `#[ignore]` on MinGW?

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -180,5 +202,5 @@ jobs:
     - name: Upload prebuilt cg_clif
       uses: actions/upload-artifact@v2
       with:
-        name: cg_clif-${{ runner.os }}
+        name: cg_clif-${{ runner.os }}-${{ matrix.env.TARGET_TRIPLE }}
```
Comment (bjorn3) [inline .github/workflows/main.yml:214]: Could you change this to just the target triple for all artifact uploads? Except for the `cg_clif-ubuntu-latest-cross-x86_64-mingw` artifact upload.

File: patches/0001-rand-Disable-rand-tests-on-mingw.patch (hunk 1 above)
Comment (bjorn3) [reply patches/0001-rand-Disable-rand-tests-on-mingw.patch:20]: And could you please use 0003-... as patch name? I try not to reuse patch numbers after patches have been dropped.

File: .github/workflows/main.yml (hunk 2 above)
Comment (afonso360) [reply .github/workflows/main.yml:214]: So `name: ${{ matrix.env.TARGET_TRIPLE }}` ? I'm not sure that's going to work, we only specify `TARGET_TRIPLE` in cross compilation scenarios, so the artifact table currently looks like this:

* `cg_clif-Linux`
* `cg_clif-Linux-cross-x86_64-mingw`
* `cg_clif-Windows-`
* `cg_clif-Windows-x86_64-pc-windows-gnu`
* `cg_clif-macOS`

Won't there be a conflict if we do that?

On a separate note, just noticed there is a trailing - on the Windows artifact.

Comment (bjorn3) [reply .github/workflows/main.yml:214]: > we only specify TARGET_TRIPLE in cross compilation scenarios.

Specifying it in non-cross compilation scenarios is fine. It defaults to the host triple, so if you set it to the host triple it doesn't affect anything.

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -112,7 +116,7 @@ jobs:
       if: matrix.env.TARGET_TRIPLE != 'x86_64-pc-windows-gnu'
       uses: actions/upload-artifact@v2
       with:
-        name: cg_clif-${{ runner.os }}
+        name: ${{ matrix.env.TARGET_TRIPLE }}
```
Comment (bjorn3) [inline .github/workflows/main.yml:119]:
```suggestion
        name: cg_clif-${{ matrix.env.TARGET_TRIPLE }}
```

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -211,5 +217,5 @@ jobs:
     - name: Upload prebuilt cg_clif
       uses: actions/upload-artifact@v2
       with:
-        name: cg_clif-${{ runner.os }}-${{ matrix.env.TARGET_TRIPLE }}
+        name: ${{ matrix.env.TARGET_TRIPLE }}
```
Comment (bjorn3) [inline .github/workflows/main.yml:220]:
```suggestion
        name: cg_clif-${{ matrix.env.TARGET_TRIPLE }}
```

**Review by bjorn3 — APPROVED** [review]
Thanks! Will merge once CI passes.

**Pull request merged**